Added documentation on how to specify the reporter for browser mocha

Lindsay-Needs-Sleep opened the pull request:
# Description of the Change
Update documentation to include usage instructions on how to specify the reporter for browser tests.

This was absent, despite hinting that it could be done.  It took quite awhile of digging through source to figure out how this could be correctly.  This information should be available.

### Alternate Designs

Thought about not including the list of possible strings, but decided I should because otherwise people will have to guess on that front as well.

### Why should this be in core?

Because it's documentation.

### Benefits

Save people time and frustration.

### Possible Drawbacks

None.

### Applicable issues
Issue #1592
<!--
* Enter any applicable Issues here.

* Mocha follows semantic versioning: http://semver.org

* Is this a breaking change (major release)?
* Is it an enhancement (minor release)?
* Is it a bug fix, or does it not impact production code (patch release)?
-->

juergba commented:
I guess the standard way to configure options, is using `mocha.setup()` as described some lines above:
`mocha.setup ( {reporter: 'html'} )`

I don't use Mocha in the browser, but I doubt the reporter lists is correct for browser usage.

Lindsay-Needs-Sleep commented:
Okay!
I'll try that out tomorrow and modify the documentation to match.

I'll try out a couple other reporters and see if they work.  All the ones I listed **are** included in mocha.js.  So it would seem that they technically **should** work (why include broken things).  I'll file a bug if they don't.

I haven't dug around too much yet.  Do you happen to know if mocha.js is exclusively for the browser?  (I guess that is pretty relevant to which reporters are loaded by mocha.js!)

Lindsay-Needs-Sleep commented:
@juergba You are absolutely right!  Thank you for the feedback!

I didn't realize this was how to do it correctly since the documentation on options for the browser is a bit sparse.  So I added to the options documentation a bit instead.

I also tested all the available builtin reporters.  They all "worked", (aka. none of them caused a failure).  Most of the time the output was legible.  I removed the list of reporters and put a link instead, (as they all basically correspond directly to the reporters list).

Lindsay-Needs-Sleep commented:
(Also not sure why the travis-ci build would fail when it is just documentation updates).  I ran `npm test`, it all passed.

outsideris commented on docs/index.md:
```diff
@@ -1391,13 +1391,13 @@ The "nyan" reporter is exactly what you might expect:
 ### Landing Strip
 
-The Landing Strip (`landing`) reporter is a gimmicky test reporter simulating a plane landing :) unicode ftw
+The Landing Strip ("landing") reporter is a gimmicky test reporter simulating a plane landing :) unicode ftw
```
Isn't it better to show them as code block rather than just quoted?

Lindsay-Needs-Sleep replied:
I think either is fine, but consistency would be best.  I went with quotes because all the rest were using quotes. 

Personally, I think it might be better to explicitly indicate that the quoted/code-block is the string name used when choosing the reporter.  Maybe something like: 

```text
reporter name = "landing"
```

Or maybe to keep changes minimal, the use the inline code-block with quotes?   
Eg. The Landing Strip (`"landing"`) reporter...  
Or is that overkill?  I kind of like the quotes because it draws extra attention and implies it must be used exactly as written. (But, I guess the code block does as well.)

I'll change it to what ever the majority/senior contributors prefer!  (So since you seem pretty senior, I'll change them all to code blocks (or whatever) if you prefer!)

Lindsay-Needs-Sleep replied:
For reporter's formatting discussion please see:
https://github.com/mochajs/mocha/issues/4054

juergba replied:
please use double quotes "landing" for consistency reasons.

juergba commented:
I agree that the browser part of our docu is suboptimal. But again I have doubts about your changes.

The option list is not the actual one of the `master` branch, the Mocha constructor I guess.
Are you sure all these options do work in the browser?

What about the `grep`, probably also `fgrep`?

> The browser may use the --grep as functionality. Append a query-string to your URL: ?grep=api.

What about `color`? `invert` is a CLI option?
Aren't the test files added manually to the browser, so file loading options do work only via CLI?

Lindsay-Needs-Sleep commented:
@juergba I just took the documentation from the code (lib/mocha.js I believe).  

But, good piont.  I did not test each option.  So I guess it's technically better to have no documentation than have documentation that indicates non-existent/non-working functionality.  I have removed the parts I did not personally test.

You were right about color.  The documentation in the code was incorrect, it is not "color" but "useColors".  I have updated that.  And I did just test out useColors out of curiorsity:

useColors: **false** (reporter: spec)
![image](https://user-images.githubusercontent.com/51773923/66192208-348c7000-e64d-11e9-8449-dbd1cc54e479.png)

useColors: **true** (reporter: spec)
![image](https://user-images.githubusercontent.com/51773923/66192443-b5e40280-e64d-11e9-9d2f-fa833cc8e4fc.png)

Lindsay-Needs-Sleep commented:
I have removed the changes to the reporters' names documentation.  That will be solely part of Pull Request #4027.

This change only includes some minor improvements to the browser documentation.

juergba commented:
Ok, let's make a last effort and finish this PR.

`color` / `useColor`:
the corresponding CLI option is `color` (with aliases `c`, `colors`), therefore the Mocha constructor should also accept an option `color` (the canonical name, no alias). There have been changes to the Mocha constructor a few weeks ago, in our `master` branch which is not published yet. Maybe you haven't tested with the `master` version of `./mocha.js`. Anyway I didn't expect `color` to have an effect in the browser.

Add the options `diff` and `inlineDiffs` to your option list, they should also work in browser.

Please also add a comment that not all CLI options are supported in the browser, with a link to the CLI options.

juergba commented on docs/index.md:
````diff
@@ -1574,6 +1574,16 @@ mocha.setup({
+```text
+reporter       {string|constructor} -  Reporter** name or constructor.
+ui             {string}             -  Interface name.
+useColors      {boolean}            -  Use color TTY output from reporter?
````
it should be `color`

juergba commented on docs/index.md:
````diff
@@ -1574,6 +1574,16 @@ mocha.setup({
+Some available options:
+
+```text
+reporter       {string|constructor} -  Reporter** name or constructor.
````
please remove those **

juergba commented on lib/mocha.js:
```diff
@@ -83,12 +82,13 @@ exports.Test = require('./test');
  * @param {number} [options.slow] - Slow threshold value.
  * @param {number|string} [options.timeout] - Timeout threshold value.
  * @param {string} [options.ui] - Interface name.
+ * @param {boolean} [options.useColors] - Use color TTY output from reporter?
```
it should be `color`

Lindsay-Needs-Sleep commented:
@juergba Sure, I can go through those changes in a couple days.  (I just have some very urgent work deadlines that I need to finish first. T.T)

Lindsay-Needs-Sleep commented:
I added `diff` and `inlineDiffs`.  I tried them out as well.  Nothing broke (after my changes).  But I didn't see anything different.  What are they supposed to do?

Lindsay-Needs-Sleep commented:
Also, changing the options to use `color` instead of `useColors` caused `this[opt](opts[opt]);` to break:

```
    mocha.setup = function(opts) {
      if (typeof opts === 'string') {
        opts = {ui: opts};
      }
      for (var opt in opts) {
        if (opts.hasOwnProperty(opt)) {
          this[opt](opts[opt]);
        }
      }
      return this;
    };
```

So I also changed:
```
Mocha.prototype.useColors = function(colors) {
```
to:
```
Mocha.prototype.color = function(colors) {
````

I had do similar changes for `diff` and `inlineDiffs`.

This caused these tests to fail:
```
1) Mocha
       constructor
         "before each" hook for "should not attempt to set timeout":
     TypeError: Cannot stub non-existent own property useInlineDiffs
      at Sandbox.stub (node_modules\sinon\lib\sinon\sandbox.js:308:19)
      at Context.<anonymous> (test\unit\mocha.spec.js:23:15)
      at callFn (lib\runnable.js:386:21)
      at Hook.Runnable.run (lib\runnable.js:373:7)
      at next (lib\runner.js:384:10)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  2) Mocha
       #hideDiff()
         should set the diff option to false when param equals true:
     TypeError: mocha.hideDiff is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:230:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  3) Mocha
       #hideDiff()
         should set the diff option to true when param equals false:
     TypeError: mocha.hideDiff is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:236:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  4) Mocha
       #hideDiff()
         should set the diff option to true when the param is undefined:
     TypeError: mocha.hideDiff is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:242:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  5) Mocha
       #hideDiff()
         should be chainable:
     TypeError: mocha.hideDiff is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:248:20)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  6) Mocha
       #useColors()
         should set the color option to true:
     TypeError: mocha.useColors is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:358:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  7) Mocha
       #useColors()
         should not create the color property:
     TypeError: mocha.useColors is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:364:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  8) Mocha
       #useColors()
         should be chainable:
     TypeError: mocha.useColors is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:370:20)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  9) Mocha
       #useInlineDiffs()
         should set the inlineDiffs option to true when param equals true:
     TypeError: mocha.useInlineDiffs is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:377:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  10) Mocha
       #useInlineDiffs()
         should set the inlineDiffs option to false when param equals false:
     TypeError: mocha.useInlineDiffs is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:383:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  11) Mocha
       #useInlineDiffs()
         should set the inlineDiffs option to false when the param is undefined:
     TypeError: mocha.useInlineDiffs is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:389:13)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:453:7
      at next (lib\runner.js:362:14)
      at Immediate.<anonymous> (lib\runner.js:421:5)

  12) Mocha
       #useInlineDiffs()
         should be chainable:
     TypeError: mocha.useInlineDiffs is not a function
      at Context.<anonymous> (test\unit\mocha.spec.js:395:20)
      at callFn (lib\runnable.js:386:21)
      at Test.Runnable.run (lib\runnable.js:373:7)
      at Runner.runTest (lib\runner.js:534:10)
      at S:\Projects\MiloProductions\00 Old and reference Repos\mocha\lib\runner.js:652:12
      at next (lib\runner.js:443:14)
```

So should I leave the prototype functions alone, and then be modifying the Mocha.setup function to not use `this[opt](opts[opt]);` and instead manually call the corresponding function for each option?

This seems way beyond the scope of my pull request intent.  I just wanted to let people know how to use custom reporters in the browser. T.T

This problem also probably extends to many of the options.  Could we create a new issue specifically for dealing with all the browser options?  (Finding out which ones actually do things, have proper names)

juergba commented:
Yes, I'm sorry. Lets squash this PR down to the absolute minimum, the reporter setup thing, and then finish it.
I will have a look at the error messages.

juergba commented:
I just have tested most of the builtin reporters, the only one which works in the browser on my laptop (Chrome/Windows10) is "HTML". All the others don't output anything.
You commented earlier that all builtin reporters were producing an output. Can you explain?
When I misspell the reporter (eg. "SPEC"), then the default reporter ("html") is used.
Otherwise with correct spelling "spec" there is no output at all.

juergba commented:
I tested some CLI options in my Chrome browser and created a draft PR #4065.
If you still want (I'm sorry, how this PR turned into a beast) you can create your own PR, copy my option list and add your own ones. But I believe only what I see with my own eyes. And I'm rather sceptical that any builtin reporter other than "html" has ever been intended for browser usage. There must be some third-party reporters though.

Lindsay-Needs-Sleep commented:
@juergba (No need to be sorry!  I was the one who started the scope creep >.<)

> I tested some CLI options in my Chrome browser and created a draft PR #4065.

Thank you so much! 

> If you still want (I'm sorry, how this PR turned into a beast) you can create your own PR, copy my option list and add your own ones.

Yes! When I have a bit more time I will be glad to contribute as well (since I have already spent some time with this, it makes sense)

> I just have tested most of the builtin reporters, the only one which works in the browser on my laptop (Chrome/Windows10) is "HTML". All the others don't output anything.
You commented earlier that all builtin reporters were producing an output. Can you explain?
When I misspell the reporter (eg. "SPEC"), then the default reporter ("html") is used.
Otherwise with correct spelling "spec" there is no output at all.

When I use:
```
    mocha.setup({
        ui: 'bdd',
        useColors: true,
        reporter: 'spec'
    });
```
This is what I see:
![image](https://user-images.githubusercontent.com/51773923/66710469-629d3e80-ed36-11e9-9b9b-25391b1ceb63.png)

Did you open the console? 

All of the reporters, except "html", output to the console instead of the html.  If you are wondering whether that is useful/appropriate for the browser, that is actually entirely why this PR was started.  I wanted the html reporter to console.log any errors.  (So I had to figure out how to supply a customized reporter to mocha.)

Another good reason for having access to reporters that log to the console is for the case of automating tests using a headless browser like phantomJS.


>Lets squash this PR down to the absolute minimum

I'll remove the extra documentation.  (Thanks you!)

Lindsay-Needs-Sleep commented:
Okay, I slimmed it down.  
If you would prefer a different format or wording, please fee free to edit it, or tell me exactly what would be preferred.  I really want to do no more than 1 more revision to this! >.<

juergba commented:
Since I have one try left ...
- If you check the "deploy/netlify preview" (under "checks") the format is wrong, one column at the left.
- I'm not happy with the console output. There are some illegible signs, and I don't want to be flooded by new issues about that. But obviously there are reasons to do so.
To finish this: we don't recommend to use CLI buitlin reporters for browser, but it's possible. So please mention this with cautious words (you can put your cat in the microwave, but don't blame us ...).

Lindsay-Needs-Sleep commented:
I updated it again.  How does it look now?  (I'll change it again if need be.)

> If you check the "deploy/netlify preview" (under "checks") the format is wrong, one column at the left.

This is very cool.  Thank you for pointing this out.  (And the mistake.)  Apparently my markdown editor preview is not very accurate.  (I'll check back when the deploy/netlify is done)

juergba commented:
The cat was meant as a picture, I was hoping you as english native would translate this into some smooth words ... meaning: you can use those CLI reporters in the browser, but they are not made/supported for that purpose.

I will finish this PR tomorrow with maybe:

> It is possible to use built-in reporters as well, but their employment in browsers is neither recommended nor supported.

Thanks!

Lindsay-Needs-Sleep commented:
> The cat was meant as a picture, I was hoping you as english native would translate this into some smooth words ... meaning: you can use those CLI reporters in the browser, but they are not made/supported for that purpose.

Haha, sorry!  I liked it, it made me laugh, and it definitely got the point across.  (But, I guess it was a bit extreme. :p)  (I wouldn’t have guessed that english is not your first language!)

Thank you for finishing this pull request though! :)
